lib/reek/spec/should_reek_of.rb

```diff
@@ -17,22 +23,42 @@ def initialize(smell_category,
 
       def matches?(actual)
         self.examiner = Examiner.new(actual, configuration: configuration)
-        self.all_smells = examiner.smells
-        all_smells.any? { |warning| warning.matches?(smell_category, smell_details) }
+
+        set_failure_messages_for_smell_type && return unless matching_smell_types?
+        set_failure_messages_for_smell_details && return unless matching_smell_details?
+
+        true
```

Comment:
After looking at it for some time I finally understood the logic behind this (two sets of messages, we set one for one failure, the other for the other failure and none + return `true` if there are no failures), but I still think it reads sub-optimally.

How about being a bit more verbose, but explicit:

```ruby
def matches?(actual)
  self.examiner = Examiner.new(actual, configuration: configuration)
  set_failure_messages
  matching_smell_types? && matching_smell_details?
end

private

def set_failure_messages
  set_failure_messages_for_smell_details unless matching_smell_details?
  set_failure_messages_for_smell_type unless matching_smell_types?
end
```